_posts/2020-09-30-hows-the-performance-of-ruby-3.0.0-preview1.md

```diff
@@ -0,0 +1,76 @@
+
+## Pitfalls
+
+First off, not every gem is ready for Rails 3. For instance, the latest version of ruby_dep (1.5.0) has a "\~>2.2" dependency on Ruby. I'm not trying to pick on it! [Tilde-dependencies are usually a really good idea](https://guides.rubygems.org/patterns/#pessimistic-version-constraint)! And now a bunch of them are going to need to change.
+
+I'm also using ancient code for this benchmark, frankly. I consider myself a pragmatist on this point, and I tried what a pragmatist would do. &lt;whispering&gt;I commented out the check for Ruby version in the local copy of Bundler 1.1.17&lt;/whispering&gt;. Um, I mean we're all respectable Rubyists here who never cut corners to get speed ratings of prerelease software.
+
+It didn't work, though. It turns out there are ***several*** changes in Ruby 3.0 that are going to require gems to upgrade a bit. But if the main problem is just the version number&hellip;
+
+I put together [a Ruby branch which started out exactly identical to the Git SHA for Ruby 3.0.0-preview1, plus a commit that rolled the reported version back to 2.8](https://github.com/noahgibbs/ruby/tree/fake_2_8). That's not as good an idea as updating all the gems to support 3.0, but it requires a lot fewer updates to ancient software and/or that I can't do for myself. Though I still have to revert a few deprecations so that old code warns instead of crashing...
+
+![A Variety of Ruby 3.0.0 Deprecations-Turned-Errors I Reverted](/blog/assets/images/ruby_30_deprec.png")
+
+And now, using version 3.0 with the serial numbers filed off, off we go!
+
+Nearly. In fact, it looks like my (badly-patched) version has an incompatibility between (un-reverted) Ruby 3.0 changes and a monkeypatch claiming to be for ActionModel::Serializer 0.8/0.9 (!). So after removing a (very small) bit of functionality in favour of returning a static string, and turning down the number of load threads a bit to avoid an occasional segfault... **Now** off we go.
+
+## Software for Testing
+
+My existing test stack seemed like the way to go: I've built this very specifically to time Ruby 3.0 against Ruby 2.0, for many years. Now that Ruby 3 exists, I can finally use it for that! It's about time!
+
+[Rails Ruby Bench](https://github.com/noahgibbs/rails_ruby_bench) runs a copy of Discourse, a common and popular Rails app to host internet forums. It's one of the biggest available "real" open-source Rails apps, making it a fine choice for "real world" benchmarking. RRB runs a set of simulated pseudorandom user requests against the running Rails app, and times how long they all take to finish. So it's a throughput test. [You can run it yourself if you like](https://engineering.appfolio.com/appfolio-engineering/2019/11/28/how-do-i-use-rails-ruby-bench), though it's a bit complex and finicky. You may be gathering that from the "Pitfalls" section above. The dark side of using real-world software is hitting real-world complexity and bugs.
```

Comment:
@noahgibbs Maybe you can include a link to https://github.com/discourse/discourse when you mention Discourse?

Reply:
Fixed, thanks!